src/tss2-tcti/tcti-libtpms.c

```diff
@@ -0,0 +1,969 @@
+ * for future features.
+ *
+ * Using tcti_libtpms_get_current_tcti(), one can retrieve the currently active
+ * libtpms TCTI instance.
+ */
+
+TPM_RESULT
+tcti_libtpms_cb_nvram_init(void)
+{
+    LOG_TRACE("tcti-libtpms callback nvram_init() called.");
+
+    return TPM_SUCCESS;
+}
+
+TPM_RESULT
+tcti_libtpms_cb_nvram_loaddata(
+    unsigned char **data,
+    uint32_t *length,
+    uint32_t tpm_number,
+    const char *name)
+{
+    LOG_TRACE("tcti-libtpms callback nvram_loaddata() called: "
+              "data=0x%" PRIxPTR ", "
+              "length=0x%" PRIxPTR ", "
+              "tpm_number=%" PRIu32 ", "
+              "name=%s",
+               (uintptr_t) data, (uintptr_t) length, tpm_number, name);
+
+    return TPM_RETRY;
+}
+
+TPM_RESULT
+tcti_libtpms_cb_nvram_storedata(
+    const unsigned char *data,
+    uint32_t length,
+    uint32_t tpm_number,
+    const char *name)
+{
+    LOG_TRACE("tcti-libtpms callback nvram_storedata() called: "
+              "data=0x%" PRIxPTR ", "
+              "length=%" PRIu32 ", "
+              "tpm_number=%" PRIu32 ", "
+              "name=%s",
+               (uintptr_t) data, length, tpm_number, name);
+
+    return TPM_SUCCESS;
+}
+
+TPM_RESULT
+tcti_libtpms_cb_nvram_deletename(
+    uint32_t tpm_number,
+    const char *name,
+    TPM_BOOL must_exist)
+{
+    LOG_TRACE("tcti-libtpms callback nvram_deletename() called: "
+              "tpm_number=%" PRIu32 ", "
+              "name=%s, "
+              "must_exist=%d",
+               tpm_number, name, must_exist);
+
+    LOG_ERROR("Not implemented");
+
+    return TPM_FAIL;
+}
+
+TPM_RESULT
+tcti_libtpms_cb_io_init(void)
+{
+    LOG_TRACE("tcti-libtpms callback io_init() called.");
+
+    return TPM_SUCCESS;
+}
+
+TPM_RESULT
+tcti_libtpms_cb_io_getlocality(
+    TPM_MODIFIER_INDICATOR *locality_modifer,
+    uint32_t tpm_number)
+{
+    TSS2_RC rc;
+    TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms;
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common;
+
+    LOG_TRACE("tcti-libtpms callback io_getlocality() called: "
+              "locality_modifer=0x%" PRIxPTR ", "
+              "tpm_number=%" PRIu32,
+               (uintptr_t) locality_modifer, tpm_number);
+
+    if (locality_modifer == NULL) {
+        return TPM_FAIL;
+    }
+
+    rc = tcti_libtpms_get_current_tcti(&tcti_libtpms);
+    if (rc != TPM2_RC_SUCCESS) {
+        return TPM_FAIL;
+    }
+    tcti_common = tcti_libtpms_down_cast(tcti_libtpms);
+    *locality_modifer = tcti_common->locality;
+
+    return TPM_SUCCESS;
+}
+
+TPM_RESULT
+tcti_libtpms_cb_io_getphysicalpresence(
+    TPM_BOOL *physical_presence,
+    uint32_t tpm_number)
+{
+    LOG_TRACE("tcti-libtpms callback io_getphysicalpresence() called: "
+              "physical_presence=0x%" PRIxPTR ", "
+              "tpm_number=%" PRIu32,
+               (uintptr_t) physical_presence, tpm_number);
+
+    LOG_ERROR("Not implemented");
+
+    return TPM_FAIL;
+}
```

Comment:
So can't we just use NULL for these not implemented callbacks in TPMLIB_RegisterCallbacks()

Reply:
Yes, we could. I decided to have them in the code anyway, so if someone is working on that code in the future, they're instantly aware of these callbacks. E.g. if you decide to implement pysical presence tests, you see that there is a callback. If you want, I'll remove the unneeded stubs.